### ISSUE TITLE
PageSize

### ISSUE THREAD

**Issue by CDrummond** [nikclayton/android-squeezer#201]
When scrolling lists (e.g. list of artists) I see lots of "Loading.." entries. This makes it hard to guess where to scroll to when looking for an item - i.e. searching for an artist in the list.

I noticed that 'PageSize' is set to 20. As a quick test, I set this to 500. Now loading this list of artists is much more responsive for me - makes navigating this list **much** nicer, with no noticeable performance hit.

I guess a PageSize of 20 was chosen to reduce the load on the server? Would it be possible to make PageSize configurable?

**Comment by kaaholst** [comment]
Comment from the Squeezebox Community Forum:

I noticed this over 3 years ago, and raised an issue (https://github.com/nikclayton/androi...zer/issues/201). I'm not sure why you use such as small pageSize, as this makes "Loading..." appear when scrolling lists. With MaterialSkin, a Javascript app, I request up to 25k at a time - but use a virtual-scroller so that only those shown on the screen are part of the document. If a JavaScript app can load that many and be responsive, a pure Java app should be able to handle even more. So, if you asked for 25k (basically all) items then only draw the visible items this would resolve your issue - and my original reported issue.

**Comment by kaaholst** [comment]
Thanks for the suggestion,

The reason for the small page size, is to reflect the number of items which fits on the screen. The intention is to fetch only the items which are displayed, or put another way, to not fetch items which we are not going to display anyway. This seems like a healthy scheme to me. The page size is not chosen to reduce the load on the server. Neither is it done to make scroll more responsive/smooth because we already use a dynamic list.

What this means to the user is that when the page loads and you need one of the top items, you can select it and proceed immediately. It is not an indication that we have fetched all the items, but since after the initial fetch we know the total number number of items, the list is fully scrollable. When you scroll down to an arbitrary position, items are fetched when they are to be displayed. Only the the 20 items close to that position (or perhaps 2 pages if the position is on or near a page boundary) will be fetched. Just wait a bit to see where we are at, and scroll up or down once the items shows. 

I don't see this makes it any harder to guess where to scroll to. You still would have to pick a position to see which items are at that position. Of course you wouldn't have to wait until the "Loading ..." disappears, at the expense of a (possibly) slower initial display of the item list. A more useful feature, in my opinion, is the alphabet shortcuts as mentioned for example here #82.

However since you are loading 25K at a time, do you have any experience how that affects the response time, versus an itemsPerResponse of 1, 20 or 200?
I realize of course that's a difficult question to answer, as it depends on the hardware on the server, number of items found for a specific request, the type of data to be fetched (how much metadata, are the data local on the server or remote on a music service, ...), the network on which the data are transferred etc.

**Comment by CDrummond** [comment]
For MaterialSkin, it does not affect the response times at all. In most cases there is **much** less than 25k - having a list this size would be unmanageable anyway. (25k was just chosen at random.) For Material I cache the results of the artists and album lists (but not the albums of artists), but even here I'm not sure it makes that much difference.

But Suqeezer's choice of 20 always seemed odd to me - and, as a user, I **hated** seeing the _Loading.._ Hence I tried setting `pageSize` to 500 and it was **much** better, with no effect on performance. Really, why should it? The JSON required for the info of 500 items is tiny, and parsing is very quick. Only getting the info for visible items seems odd, because as soon as you scroll you no longer have the info - hence _Loading..._

Not that it proves much, but I've never had a user complain on the slimdevices forum nor github that Material's list fetching is slow.

**Comment by kaaholst** [comment]
Thanks for the reply, that's good to know.

Since most lists are limited in size, for the majority of cases the page size won't matter. Some list are not limited though, like artists, albums, tracks, radio stations and perhaps others. For those cases, to protect from memory issues and slow load times, I prefer to use a paging scheme. After all this has worked for years and no one, except you, seems to notice it.

Based on your experience though I will increase the page size, and also remove the fetching of the first element, to workaround some server side issues where the page size is confused with the number of items available.

**Comment by kaaholst** [comment]
Just run some tests for request for first page of a search for tracks returning 254 results.
```
Page size Old NAS New NAS     PC
        1    2,14    0,53   0,16
       20    4,35    0,53   0,16
       50    8,15    0,55   0,19
      200   26,56    0,76   0,25
```

**Comment by kaaholst** [comment]
This is fixed in Squeezer version 2.2.1 and newer, closing.
